Print NFA to mata format (#254)
VeriFIT/mata: AUTOMATAFORMAT.md

> # The file format for automata (.mata format)

Reviewer: I suggest changing the tile so that it is immediately visible that this is Mata format, like so:
```suggestion
# The Mata file format for automata (.mata format)
```

* Besides white spaces and the end of line, the following are **characters with special meaning**: `&`,`|`,`!`,`@`,`(`,`)`,`%`,`"`,`\`,`#` `[`,`]`,`a`,`q`,`n`,`t`,`f`.
  * `&`,`|`,`!`,`(`,`)` are logical operators used in Boolean formulas.
  * `[`,`]`,`-`,`^` enclose character classes, such as `[abcd-h]`, `[^abcd-h]`.
  * `@` opens the line with a section name and is used in transducer alphabet tokens of the form `x@a1`, `y@[10-55]`,...
  * `%` opens a key-value line.
  * `"` strings containing white spaces and special characters can be written in between `"`. The special characters lose their special meaning. The characters `"` and `\` inside a string must be escaped, i.e. `\"` and `\\`. 
  * `\` for escaping characters. `\` is also used to concatenate lines.
  * `a`,`q`,`n`,`t`,`f` are token type specifiers (alphabet, state, node, attribute, formula). 
  * `#` starts a comment line.

> * There are **words with a special meaning**: `\true`,`\false`, `\min`, `\max`. First two represents true/false in Boolean formulas.

Reviewer: Slightly better wording.
```suggestion
* There are **words with a special meaning**: `\true`,`\false`, `\min`, `\max`. The first two represent true/false in Boolean formulas.
```

* The parser may recognise the special symbols as standalone tokens (does not apply to `q`,`a`,`n`,`t`,`f`,`#`) only if they appear in the special context where their special meaning applies, otherwise they should be treated as normal symbols.

We categorise automata according to their **transition type**, that is, the structure of their transitions, and by their **alphabet type**, i.e., how alphabet symbols on transitions are represented. It is expected that different parameters would be parsed into different data structures. These parameters should determine the name of the automata section.

## Alphabet type
We categorise automata by the **alphabet type**, i.e., the representation of symbols on transitions. We will consider propositional formulae (over bit vectors), explicit symbols, intervals of numbers or unicode or ascii, or user-defined alphabet.
* **Explicit**: Just plain symbols. May be given implicitly or enumerated, or one can have some predefined alphabet, numbers, ascii, utf.
* **Bitvector**: Propositional variables, the syntax is the same as for explicit symbols.

> * **Character class**: Character classes, i.e. sets of symbols as used in normal regular expressions, are of the form `[bla]` or `[^bla]` where bla is a sequence of symbols and intervals of symbols such as `a-z`, for instance `abcd-hij-z` denotes all lower case alphabet symbols. The `^` in front complements the entire class. The character `-` in sincluded as `\-`, analogously `\` and `^`. The special keywords `\min` and `\max` denote the first repsective the last letter the alphabet.

Reviewer: ```suggestion
* **Character class**: Character classes, i.e. sets of symbols as used in normal regular expressions, are of the form `[bla]` or `[^bla]` where `bla` is a sequence of symbols and intervals of symbols such as `a-z`, for instance `abcd-hij-z` denotes all lower case alphabet symbols. The `^` in front complements the entire class. The character `-` in sincluded as `\-`, analogously `\` and `^`. The special keywords `\min` and `\max` denote the first repsective the last letter the alphabet.
```

## Automata by their transition type

### NFA (Nondeterministic finite automata)
* The transitions are triples consisting of a source state, transition formula, and target state. Note that a single explicit symbol or a single interval is also a transition formula. 
* Both symbol literals and states are determined positionally. Markers or enumeration can also be used.
* Initial (final) states are defined by key-value line `%Initial"` (`%Final`) followed by an enumeration of states.

### AFA (Alternating finite automata)

> * Transition lines are of the form `"<State> <Formula>"` representing the source state and the transition formula - a Booelan formula over symbol literals, states, and nodes.

Reviewer: ```suggestion
* Transition lines are of the form `"<State> <Formula>"` representing the source state and the transition formula - a Boolean formula over symbol literals, states, and nodes.
```

* Several lines starting with the same state mean the disjunciton of the formulae. No line for the state means false. 
* Initial (final) states are defined by key-value line `%Initial <Formula>"` (`"%Final <Formula>"`) where the Boolean formula is given over states.

> The transition and alphabet type is encoded in the name of a section, e.g., `@NFA-explicit` denotes a nondeterministic finite automaton over explicit alphabet is given.

Reviewer: ```suggestion
The transition and alphabet type is encoded in the name of a section, e.g., `@NFA-explicit` denotes that a nondeterministic finite automaton over an explicit alphabet is given.
```
